### ISSUE TITLE
Switching inliners from the command line

### ISSUE THREAD

**Issue by efferifick** [eclipse/omr#3956]
Hi, 

I'd like to add infrastructure to allow people to easily try out non-default inliners - eg allow experimentation. This post is mostly trying to encourage a discussion on the different ways one can switch inliners from command line options.

There are several ways I can imagine this to happen:

1. Replacing the reference to `create` method in the `_opts` array.
```C++
_opts[OMR::inlining] = condition ?
new (comp->allocator()) TR::OptimizationManager(self(), TR_TrivialInliner::create, OMR::inlining) :
new (comp->allocator()) TR::OptimizationManager(self(), TR_ExperimentalInliner::create, OMR::inlining) ;
```

2. Another option: Have an inliner wrapper which defaults to whatever class is passed by a template and on a command line option switch to the experimental inliner.
```C+_
_opts[OMR::inlining] = 
TR::OptimizationManger(self(), TR_SelectInliner<TR_Inliner>::create, OMR::inlining);
_opts[OMR::trivialInlining] =
new (comp->allocator()) TR::OptimizationManager(self(), TR_SelectInliner<TR_TrivialInliner>::create, OMR::trivialInlining);
```
(or some other way of passing the default inlining strategy that is not using templates). But the goal is to have the `TR_SelectInliner` class  do either the default inlining pass (which would vary depending on the instance of `TR_SelectInliner`) or the (in this example) `TR_ExperimentalInliner`.

3. Have an optimization pass that is a wrapper around the inlining strategies and that multiplexes based on flags or hotness. I've been told this is not a good way since now the optimization pass is selected inside the optimization as opposed to the optimization plan, but I added it here just for discussion around it as well. (It also doesn't consider the targeted inlining and would need a way to determine when to run that inliner.)
For example:
```C++
int32_t
OMR::SelectInliner::perform() {
   if (this->comp()->getOption(TR_EnableExperimentalInliner))
   {
      return this->performExperimentalInliner();
   }

   if (this->comp()->getMethodHotness() >= warm)
   {
      return this->performMultiTargetInliner();
   }

   return this->performTrivialInliner();
}

int32_t
OMR::SelectInliner::performTrivialInliner()
   {
   TR_TrivialInliner *inliner = (TR_TrivialInliner *)TR_TrivialInliner::create(this->_manager);
   return inliner->perform();
   }

int32_t
OMR::SelectInliner::performMultiTargetInliner()
   {
   TR_Inliner *inliner = (TR_Inliner *)TR_Inliner::create(this->_manager);
   return inliner->perform();
   }

int32_t
OMR::SelectInliner::performExperimentalInliner()
   {
   TR_VerboseLog::vlogAcquire();
   TR_VerboseLog::writeLine(TR_Vlog_SIP, "Experimental Inliner");
   TR_VerboseLog::vlogRelease();
   return 0;
}
```

**Comment by andrewcraik** [comment]
@0xdaryl this is a good one for the architecture meeting to discuss how to facilitate this kind of extension since there are a number of ways to go. I'm not a fan of moving the inliner choice at different opt levels out of the opt strategy and into the select inliner implementation. Quite how to wrap the inliner is a point worthy of more general debate IMO.

**Comment by 0xdaryl** [comment]
@efferifick, are you OK to discuss this Thursday May 6 @ 2pm EDT at the Compiler Architecture meeting?  We are in need of a topic this week.

**Comment by efferifick** [comment]
@0xdaryl yeah! Just let me know how to attend the meeting, I think Andrew mentioned WebEx?

**Comment by andrewcraik** [comment]
The architecture meeting was inclined towards option 2 for fitting with the existing design. @efferifick has agreed to create a prototype and to create a pull request where we can discuss this further. If the design needs further review we can put it on the agenda again for the next meeting.

**Comment by efferifick** [comment]
Hi, sorry for all the amends. I had to change e-mails and make sure to sign it with the correct e-mail to adhere to the contribution guidelines.
